File: ToDo.Server/Services/TasksService.cs
using ToDo.Server.Models;
using ToDo.Server.Repositories.Interfaces;
using TaskStatus = ToDo.Server.Models.TaskStatus;

namespace ToDo.Server.Services;

public class TasksService
{
    private readonly ITasksRepository _tasksRepository;

    public TasksService(ITasksRepository tasksRepository)
    {
        _tasksRepository = tasksRepository;
    }

    public async Task<List<Tasks>> GetTasksByUser(Guid userId)
    {
        var tasksByUser = await _tasksRepository.GetTasksByUser(userId);

        return tasksByUser;
    }

    public async Task<Tasks> GetTaskById(Guid taskId)
    {
        var task = await _tasksRepository.GetTaskById(taskId) ?? throw new Exception("Task not found.");

        return task;
    }

    public async Task<Tasks> CreateTask(Tasks task)
    {
        return await _tasksRepository.CreateTask(task);
    }

    public async Task<Tasks> UpdateTask(Guid taskId, Tasks updateTask)
    {
        var currentTask = await GetTaskById(taskId);

        currentTask.Title = updateTask.Title;
        currentTask.Description = updateTask.Description;
        currentTask.Status = updateTask.Status;
        currentTask.DueDate = updateTask.DueDate;

        await _tasksRepository.UpdateTask(currentTask);
        return currentTask;
    }

    public async Task<Tasks> UpdateTaskStatus(Guid taskId, TaskStatus taskStatus)
    {
        var task = await _tasksRepository.GetTaskById(taskId) ?? throw new Exception("Task not found.");

        var updatedTask = await _tasksRepository.UpdateTaskStatus(taskId, taskStatus) ?? throw new Exception("Failed to update task status.");

        return updatedTask;
    }

    public async Task DeleteTask(Guid taskId)
    {
        await _tasksRepository.DeleteTask(taskId);
    }
}

File: todo.client/src/services/tasksService.ts
import { ITask, TasksStatus } from "../types/task";
import apiRequest from "./helpers/apiService";

export const GetTasksByUser = async (userId: string) => apiRequest("GET", `tasks/byuser/${userId}`, undefined, true);

export const GetTasksById = async (taskId: string) => apiRequest("GET", `tasks/${taskId}`, undefined, true);

export const CreateTask = async (newTask: ITask) => apiRequest("POST", "tasks", newTask, true);

export const UpdateTask = async (taskId: string, newTaskData: Partial<ITask>) => apiRequest("PUT", `tasks/${taskId}`, newTaskData, true);

export const UpdateTaskStatus = async (taskId: string, status: TasksStatus) => apiRequest("PATCH", `tasks/${taskId}/status`, status, true);

export const DeleteTask = async (taskId: string) => apiRequest("DELETE", `tasks/${taskId}`, undefined, true);